Source/Input.cpp

```diff
@@ -0,0 +1,182 @@
+	// First, let's find the address of get_keyboard_button_state(). This takes two parameters: a pointer to the input device (which stored in RCX) and a byte
+	//   that represents the key being pressed (stored in EDX). The actual key byte generally changes between versions, but there's always a call to check some
+	//   sort of control marker at 0x0.
+	std::vector<byte> keyboardFunctionSearchBytes = {
+		0x33, 0xD2,				// XOR EDX,EDX <- clears the key parameter
+		0xC1, 0xE8, 0x02,		// (part of a previous operation)
+		0x48, 0x8B, 0xCF,		// MOV RCX,RDI <- move input_devices into RCX
+		0x24, 0x01,				// (part of a previous operation)
+		0x88, 0x43, 0x29,		// (part of a previous operation. this is brittle because it assumes input_devices->cursor_released is at index+0x29
+		0xE8					// CALL ...
+	};
+
+	int keyboardFunctionIndex = Utilities::findSequence<byte>(overlayFunctionBody, keyboardFunctionSearchBytes);
+	if (keyboardFunctionIndex == -1) {
+		throw "Unable to find index of Input_Devices::get_keyboard_button_state pointer in Overlay::begin().";
+		return false;
+	}
+
+	keyboardFunctionIndex += keyboardFunctionSearchBytes.size(); // Advance past the search sequence to start of pointer.
+
+	// Since referencing global values is relative, we must compute the absolute value ourselves based on the address of the next operation.
+	uint32_t keyboardFunctionJumpLength = *(uint32_t*)(&overlayFunctionBody[keyboardFunctionIndex]);
+	uint32_t keyboardFunctionOffset = keyboardFunctionJumpLength + overlayFunctionOffset + keyboardFunctionIndex + 0x4;
+
+	// Next, read the body of Input_Devices::get_keyboard_button_state().
+	const SIZE_T keyboardButtonFunctionSize = 0x20;
+	std::vector<byte> keyboardButtonFunctionBody(keyboardButtonFunctionSize);
+	if (!memory->ReadRelative(reinterpret_cast<void*>(keyboardFunctionOffset), &keyboardButtonFunctionBody[0], keyboardButtonFunctionSize)) {
+		throw "Unable to read Input_Devices::get_keyboard_button_state into memory.";
+		return false;
+	}
+
+	// Verify that we have indeed jumped to the right function by checking to see if the first two bytes are a MOV operation.
+	if (!(keyboardButtonFunctionBody[0] == 0x48 && keyboardButtonFunctionBody[1] == 0x8b)) {
+		throw "Memory at pointer indicated by Input_Devices::get_keyboard_button_state did not start with the expected MOV operation.";
+	}
+
+	// InputDevices::get_keyboard_button_state() is simple: it retrieves the pointer to the keyboard from input_devices, then calls Keyboard::get_button_state().
+	//   In order to retrieve the pointer ourselves, we need to find the offset within input_devices that represents the keyboard pointer, which we can do by
+	//   reading the offset parameter passed to the MOV operation.
+	// NOTE: The actual MOV call is slightly inconsistent between builds and may use either 1 or 4 bytes for its offset, so we need to detect that.
```

Comment:
The MOV may be a different opcode but I bet the offset is stable.

By the way, Memory is generally designed for you to provide a series of offsets, ala a cheat engine pointer path (or, originally, a livesplit variable watch).

E.g. the "WritePanelData" function is starting from `witness64_d3d11.globals` (a static address), then adds `0x18` (`globals->entity_manager.entity_array`), then `id * 8` (`globals->entity_manager.entity_array[id]`), etc etc.

Reply:
(Additionally, if you compute out all the pointer paths by hand, you run the risk of them being incorrect after a level load, or anything else that causes the game to reallocate the structure.)

Reply:
Thanks for pointing that out! I've reworked this to use Memory::ComputeOffset and am now using static values for those offsets.